csrc/kernel_cache.cpp

```diff
@@ -1598,35 +1601,92 @@ std::optional<FusionKernelRuntime::HeuristicsPtr> FusionKernelRuntime::
         const KernelArgumentHolder& args,
         std::optional<PrimDataType> forced_index_type) {
   FUSER_PERF_SCOPE("FusionKernelRuntime::getMaybeHeuristicsFor");
-  auto complete_fusion = segmented_fusion_->completeFusion();
-  precomputed_values_->bindInputs(args);
-  precomputed_values_->evaluate();
-  SchedulerRuntimeInfo runtime_info(
-      complete_fusion,
-      args,
-      precomputed_values_.get(),
-      all_tvs_,
-      forced_index_type);
-
-  std::optional<FusionKernelRuntime::HeuristicsPtr> ret;
-  ret = std::make_unique<FusionHeuristics>();
-  size_t total_groups = segmented_fusion_->groups().size();
-  for (const auto group_index : c10::irange(total_groups)) {
-    auto group = segmented_fusion_->groups()[group_index];
-
-    auto maybe_scheduler_entry = group->getMaybeSchedulerEntry(runtime_info);
-    if (!maybe_scheduler_entry.has_value()) {
-      return std::nullopt;
+
+  // The runtime group run order is different from the segmented_fusion group
+  // order. Instead of using FusionHeuristics::emplaceBack, we initialize
+  // FusionHeuristics with the desired number of groups.
+  const int64_t num_groups = (int64_t)runtime_workspace_.group_run_order.size();
+  FusionKernelRuntime::HeuristicsPtr heuristics =
+      std::make_unique<FusionHeuristics>(num_groups);
+
+  // Store metadata copy of arguments for ArgumentManager
+  KernelArgumentHolder args_metadata = copyMetadataArg(args);
+
+  // ArgumentManager manipulates the KernelArgumentHolder argument.
+  // We make another metadata copy for the PrecomputedValues.
+  KernelArgumentHolder complete_fusion_metadata_args(args_metadata);
+
+  ArgumentManager args_manager(
+      args_metadata, runtime_workspace_, segmented_fusion_->inputs());
+
+  // Follow group run order
+  for (int64_t group_id : c10::irange(num_groups)) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(group_id);
+
+    // Create fusion for this segmented group
+    Fusion* fusion_to_run = group_to_run->getFusion();
+    NVF_ERROR(fusion_to_run != nullptr);
+    FusionGuard fg(fusion_to_run);
+
+    // Get input arguments for SchedulerRuntimeInfo
+    KernelArgumentHolder group_runtime_inputs;
+    for (auto input : group_to_run->inputs()) {
+      group_runtime_inputs.push(*args_manager.checkTensorMap(input));
     }
-    auto scheduler_entry = std::move(maybe_scheduler_entry.value());
-    if (!scheduler_entry->sameAs(
-            heuristics_->heuristicsList()[group_index].get())) {
-      return std::nullopt;
+
+    // Create PrecomputedValues initialized with original fusion inputs
+    auto evaluator_precomputed_values =
+        std::make_unique<PrecomputedValues>(fusion_to_run);
+    evaluator_precomputed_values->bindInputs(group_runtime_inputs);
+    evaluator_precomputed_values->bindValues(
+        group_to_run->getCompleteFusionInputs(), complete_fusion_metadata_args);
```

Comment:
evaluator_precomputed_values now describes the cloned fusion segment. Why do we bind the complete fusion's inputs?

Reply:
It is a rare corner case, but there can still be dependencies on the original fusion's arguments.

To reproduce:
1. Disable this line
2. Run `python test/python/test_python_frontend.py -k test_issue1277`